rust/src/workflow.rs

```diff
@@ -0,0 +1,529 @@
+        };
+        unsafe { Workflow::from_raw(ptr::NonNull::new(result).unwrap()) }
+    }
+
+    /// Make a new Workflow, copying all Activities and the execution strategy.
+    ///
+    /// * `name` - the name for the new Workflow
+    /// * `activity` - if specified, perform the clone operation using
+    /// ``activity`` as the root
+    pub fn new_from_copy<S: BnStrCompatible, A: IntoActivityName>(
+        &self,
+        name: S,
+        activity: A,
+    ) -> Workflow {
+        let name = name.into_bytes_with_nul();
+        let activity = activity.activity_name();
+        unsafe {
+            Self::from_raw(
+                ptr::NonNull::new(BNWorkflowClone(
+                    self.as_raw(),
+                    name.as_ref().as_ptr() as *const ffi::c_char,
+                    activity.as_ref().as_ptr() as *const ffi::c_char,
+                ))
+                .unwrap(),
+            )
+        }
+    }
+
+    /// List of all Workflows
+    pub fn list() -> Array<Workflow> {
+        let mut count = 0;
+        let result = unsafe { BNGetWorkflowList(&mut count) };
+        assert!(!result.is_null());
+        unsafe { Array::new(result, count, ()) }
+    }
+
+    pub fn name(&self) -> BnString {
+        let result = unsafe { BNGetWorkflowName(self.as_raw()) };
+        assert!(!result.is_null());
+        unsafe { BnString::from_raw(result) }
+    }
+
+    /// Register this Workflow, making it immutable and available for use.
+    ///
+    /// * `configuration` - a JSON representation of the workflow configuration
+    pub fn register<S: BnStrCompatible>(&self, config: S) -> Result<(), ()> {
+        let config = config.into_bytes_with_nul();
+        if unsafe {
+            BNRegisterWorkflow(
+                self.as_raw(),
+                config.as_ref().as_ptr() as *const ffi::c_char,
+            )
+        } {
+            Ok(())
+        } else {
+            Err(())
+        }
+    }
+
+    /// Register an Activity with this Workflow.
+    ///
+    /// * `activity` - the Activity to register
+    /// * `subactivities` - the list of Activities to assign
+    pub fn register_activity<I>(&self, activify: &Activity, subactivities: I) -> Activity
+    where
+        I: IntoIterator,
+        I::Item: IntoActivityName,
+    {
+        let subactivities_raw: Vec<BnString> = subactivities
+            .into_iter()
+            .map(|x| x.activity_name())
+            .collect();
+        let mut subactivities_ptr: Vec<*const _> =
+            subactivities_raw.iter().map(|x| x.as_ptr()).collect();
+        let result = unsafe {
+            BNWorkflowRegisterActivity(
+                self.as_raw(),
+                activify.as_raw(),
+                subactivities_ptr.as_mut_ptr(),
+                subactivities_ptr.len(),
+            )
+        };
+        unsafe { Activity::from_raw(ptr::NonNull::new(result).unwrap()) }
+    }
```

Comment:
This does not work, given:
```rs
    let work = Workflow::new("testWorkflow");
    let activity = Activity::new_with_action("test", |ctx| {});
    work.register_activity::<Vec<String>>(&activity, vec![]);
    work.register(
        r#"{ 
            "title" : "Rust workflow",
            "description" : "",
            "capabilities" : []
        }"#,
    )
    .expect("workflow is registered");
    work.insert("core.function.translateTailCalls", ["test"]);
```
I don't know *why* but it fails in `Workflow::Validate` but it shouldn't fail if there is no subactivity.